nova_plugin/server.py

```diff
@@ -1134,53 +1131,59 @@ def _get_keypair_name_by_id(nova_client, key_name):
-    # isn't possible on Openstack)
-    new_nic_nodes = \
-        [node_instance_id for node_instance_id, runtime_props in
-         ctx.capabilities.get_all().iteritems() if runtime_props.get(
-             OPENSTACK_TYPE_PROPERTY) in (PORT_OPENSTACK_TYPE,
-                                          NETWORK_OPENSTACK_TYPE) and
-         not is_external_resource_by_properties(
-             _get_properties_by_node_instance_id(node_instance_id))]
-    if new_nic_nodes:
-        raise NonRecoverableError(
-            "Can't connect new port and/or network nodes to a server node "
-            "with '{0}'=True".format(USE_EXTERNAL_RESOURCE_PROPERTY))
-
-    # validate all expected connected networks and ports are indeed already
-    # connected to the server. note that additional networks (e.g. the
-    # management network) may be connected as well with no error raised
-    if not network_ids and not port_ids:
-        return
+def _validate_external_server_nics(external_server, network_ids, port_ids):
+    # check currently attached ports
+    interfaces = external_server.interface_list()
+    attached_ports = set([interface.port_id for interface in interfaces])
+    attached_nets = set([interface.net_id for interface in interfaces])
 
-    server_id = get_openstack_id(ctx)
-    connected_ports = neutron_client.list_ports(device_id=server_id)['ports']
-
-    # not counting networks connected by a connected port since allegedly
-    # the connection should be on a separate port
-    connected_ports_networks = {port['network_id'] for port in
-                                connected_ports if port['id'] not in port_ids}
-    connected_ports_ids = {port['id'] for port in
-                           connected_ports}
-    disconnected_networks = [network_id for network_id in network_ids if
-                             network_id not in connected_ports_networks]
-    disconnected_ports = [port_id for port_id in port_ids if port_id not
-                          in connected_ports_ids]
-    if disconnected_networks or disconnected_ports:
+    already_attached_ports = [port_id for port_id in attached_ports
+                              if port_id in port_ids]
+    already_attached_nets = [net_id for net_id in attached_nets
+                             if net_id in network_ids]
+    if already_attached_ports or already_attached_nets:
         raise NonRecoverableError(
-            'Expected external resources to be connected to external server {'
-            '0}: Networks - {1}; Ports - {2}'.format(server_id,
-                                                     disconnected_networks,
-                                                     disconnected_ports))
+            'Several ports/networks already connected to external server '
+            '{0}: Networks - {1}; Ports - {2}'
+            .format(external_server.human_id,
+                    already_attached_nets,
+                    already_attached_ports))
```

Comment:
can't this just be a warning?

Reply:
In case warning, on uninstall we will disconnect preconnected port from vm, so looks like better to raise exception and user will recheck what ports should be connected. 

But for "backward compatibility" we can ignore such ports/networks - and uninstall  will detach ports.